modal/shared_volume.py

```diff
@@ -64,16 +64,29 @@ async def read_file(self, path: str) -> AsyncIterator[bytes]:
             async for data in blob_iter(response.data_blob_id, self._client.stub):
                 yield data
 
+    async def iterdir(self, path: str) -> AsyncIterator[api_pb2.SharedVolumeListFilesEntry]:
+        """Iterate over all files in a directory in the shared volume.
+
+        * Passing a directory path lists all files in the directory (names are relative to the directory)
+        * Passing a file path returns a list containing only that file's listing description
+        * Passing a glob path (including at least one * or ** sequence) returns all files matching that glob path (using absolute paths)
+        """
+        req = api_pb2.SharedVolumeListFilesRequest(shared_volume_id=self._object_id, path=path)
+        async for batch in unary_stream(self._client.stub.SharedVolumeListFilesStream, req):
+            for entry in batch.entries:
+                yield entry
+
     async def listdir(self, path: str) -> List[api_pb2.SharedVolumeListFilesEntry]:
         """List all files in a directory in the shared volume.
 
         * Passing a directory path lists all files in the directory (names are relative to the directory)
         * Passing a file path returns a list containing only that file's listing description
         * Passing a glob path (including at least one * or ** sequence) returns all files matching that glob path (using absolute paths)
         """
-        req = api_pb2.SharedVolumeListFilesRequest(shared_volume_id=self._object_id, path=path)
-        response = await retry_transient_errors(self._client.stub.SharedVolumeListFiles, req)
-        return list(response.entries)
+        entries = []
+        async for entry in self.iterdir(path):
+            entries.append(entry)
+        return entries
```

Comment:
you can use a list comprehension for async stuff too. can collapse these 4 lines into just `return [entry async for entry in self.iterdir(path)]`